File: riveredge-backend/src/apps/kuaiplm/schemas/engineering_change.py
"""工程变更 Schema（R-04 / ECN）"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class EcnMaterialLineIn(BaseModel):
    material_id: Optional[int] = None
    material_code: str = Field(..., min_length=1, max_length=80)
    material_name: str = Field(..., min_length=1, max_length=200)
    before_desc: Optional[str] = Field(None, max_length=500)
    after_desc: Optional[str] = Field(None, max_length=500)
    stock_qty: Optional[Decimal] = None
    unit_price: Optional[Decimal] = None
    cost_amount: Optional[Decimal] = None
    disposition: Optional[str] = Field(None, max_length=20)
    owner_user_id: Optional[int] = None
    owner_user_name: Optional[str] = Field(None, max_length=100)
    remarks: Optional[str] = Field(None, max_length=500)
    extension_payload: Optional[dict] = None


class EcnMaterialLineOut(EcnMaterialLineIn):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: UUID
    line_no: int


class EcnSignoffOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: UUID
    dept_code: str
    dept_name: str
    sort_order: int
    status: str
    result: Optional[str] = None
    signer_id: Optional[int] = None
    signer_name: Optional[str] = None
    signed_at: Optional[datetime] = None
    notes: Optional[str] = None


class EngineeringChangeCreate(BaseModel):
    change_kind: str = Field(..., description="material/process/drawing/doc_template/other")
    title: str = Field(..., min_length=1, max_length=200)
    project_id: Optional[int] = None
    ecn_code: Optional[str] = Field(None, max_length=50)
    change_reason: Optional[str] = None
    remarks: Optional[str] = None
    extension_payload: Optional[dict] = None
    materials: List[EcnMaterialLineIn] = Field(default_factory=list)


class EngineeringChangeUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    change_reason: Optional[str] = None
    remarks: Optional[str] = None
    extension_payload: Optional[dict] = None
    materials: Optional[List[EcnMaterialLineIn]] = None


class EngineeringChangeErpAudit(BaseModel):
    erp_ecn_no: str = Field(..., min_length=1, max_length=80)
    result: str = Field(..., description="pass/fail")
    notes: Optional[str] = None


class EngineeringChangeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    uuid: UUID
    id: int
    ecn_code: str
    project_id: Optional[int] = None
    project_code: Optional[str] = None
    project_name: Optional[str] = None
    change_kind: str
    title: str
    change_reason: Optional[str] = None
    status: str
    erp_ecn_no: Optional[str] = None
    erp_audit_status: Optional[str] = None
    erp_audit_notes: Optional[str] = None
    remarks: Optional[str] = None
    extension_payload: Optional[dict] = None
    submitted_at: Optional[datetime] = None
    approved_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None
    materials: List[EcnMaterialLineOut] = Field(default_factory=list)
    signoffs: List[EcnSignoffOut] = Field(default_factory=list)


class EngineeringChangeListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    uuid: UUID
    id: int
    ecn_code: str
    project_id: Optional[int] = None
    project_code: Optional[str] = None
    project_name: Optional[str] = None
    change_kind: str
    title: str
    status: str
    erp_ecn_no: Optional[str] = None
    erp_audit_status: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None


class EngineeringChangeListResponse(BaseModel):
    items: List[EngineeringChangeListItem]
    total: int


class EcnFormProfile(BaseModel):
    """ECN 表单/明细列扩展 profile（通用默认或行业包覆盖）。"""

    industry_profile_enabled: bool = Field(
        False, description="是否已启用行业包 profile（false 时前端保持通用 UI）"
    )
    field_labels: Dict[str, str] = Field(default_factory=dict)
    material_line_columns: List[Dict[str, Any]] = Field(default_factory=list)
    signoff_depts: List[Dict[str, Any]] = Field(default_factory=list)
    header_option_flags: List[Dict[str, Any]] = Field(default_factory=list)
    header_fields: List[Dict[str, Any]] = Field(default_factory=list)
    entry_sources: List[Dict[str, Any]] = Field(default_factory=list)
    validation_rules: List[Dict[str, Any]] = Field(default_factory=list)

File: riveredge-frontend/src/apps/kuaiplm/services/engineering-change.ts
/**
 * 工程变更（R-04 / ECN）API
 */

import { api } from '../../../services/api';

const BASE = '/apps/kuaiplm/engineering-changes';

export type EcnStatus =
  | 'draft'
  | 'pending'
  | 'approved'
  | 'erp_pending'
  | 'erp_failed'
  | 'closed';

export type EcnChangeKind = 'material' | 'process' | 'drawing' | 'doc_template' | 'other';

export interface EcnMaterialLine {
  id?: number;
  uuid?: string;
  line_no?: number;
  material_id?: number | null;
  material_code: string;
  material_name: string;
  before_desc?: string | null;
  after_desc?: string | null;
  stock_qty?: number | string | null;
  unit_price?: number | string | null;
  cost_amount?: number | string | null;
  disposition?: string | null;
  owner_user_id?: number | null;
  owner_user_name?: string | null;
  remarks?: string | null;
  extension_payload?: Record<string, unknown> | null;
}

export interface EcnFormProfileColumn {
  key: string;
  label: string;
  sort?: number;
  required?: boolean;
  width?: number;
  type?: string;
}

export interface EcnFormProfileFlag {
  key: string;
  label: string;
  sort?: number;
  type?: string;
}

export interface EcnFormProfileHeaderField {
  key: string;
  label: string;
  sort?: number;
  type?: string;
}

export interface EcnFormProfile {
  industry_profile_enabled?: boolean;
  field_labels: Record<string, string>;
  material_line_columns: EcnFormProfileColumn[];
  signoff_depts: Array<{ code: string; label: string; sort?: number }>;
  header_option_flags: EcnFormProfileFlag[];
  header_fields?: EcnFormProfileHeaderField[];
  entry_sources: Array<{ code: string; label: string; sort?: number }>;
  validation_rules: Array<Record<string, unknown>>;
}

export interface EcnSignoff {
  id: number;
  uuid: string;
  dept_code: string;
  dept_name: string;
  sort_order: number;
  status: string;
  result?: string | null;
  signer_name?: string | null;
  signed_at?: string | null;
  notes?: string | null;
}

export interface EngineeringChange {
  uuid: string;
  id: number;
  ecn_code: string;
  project_id?: number | null;
  project_code?: string | null;
  project_name?: string | null;
  change_kind: EcnChangeKind | string;
  title: string;
  change_reason?: string | null;
  status: EcnStatus | string;
  erp_ecn_no?: string | null;
  erp_audit_status?: string | null;
  erp_audit_notes?: string | null;
  remarks?: string | null;
  extension_payload?: Record<string, unknown> | null;
  submitted_at?: string | null;
  approved_at?: string | null;
  closed_at?: string | null;
  created_at: string;
  updated_at: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
  materials?: EcnMaterialLine[];
  signoffs?: EcnSignoff[];
}

export interface EngineeringChangeListResponse {
  items: EngineeringChange[];
  total: number;
}

export const engineeringChangeApi = {
  formProfile: async () =>
    (await api.get(`${BASE}/meta/form-profile`)) as EcnFormProfile,
  list: async (params: Record<string, unknown>) =>
    (await api.get(BASE, { params })) as EngineeringChangeListResponse,
  get: async (id: number) => (await api.get(`${BASE}/${id}`)) as EngineeringChange,
  create: async (data: Record<string, unknown>) =>
    (await api.post(BASE, data)) as EngineeringChange,
  update: async (id: number, data: Record<string, unknown>) =>
    (await api.put(`${BASE}/${id}`, data)) as EngineeringChange,
  remove: async (id: number) => {
    await api.delete(`${BASE}/${id}`);
  },
  submit: async (id: number) =>
    (await api.post(`${BASE}/${id}/submit`)) as EngineeringChange,
  approve: async (id: number) =>
    (await api.post(`${BASE}/${id}/approve`)) as EngineeringChange,
  reject: async (id: number) =>
    (await api.post(`${BASE}/${id}/reject`)) as EngineeringChange,
  erpAudit: async (id: number, data: { erp_ecn_no: string; result: string; notes?: string }) =>
    (await api.post(`${BASE}/${id}/erp-audit`, data)) as EngineeringChange,
};
